File: src/me.ts
/**
 * `client.me()` — who the caller is, and what answered.
 *
 * One GET, no envelope. The response body is **flat** — `{tenant, subject,
 * account, role, versions?}` — because `/api/me` is an alias of the host's
 * existing identity handler rather than a second one (`docs/endpoints.md` §1,
 * D15). There is therefore nothing for `unwrap` to do here, and this is the one
 * read operation in the package that does not call it.
 *
 * **`status: "planned"`** in `endpoints.json`: the identity half is served today
 * at a different path, and `/api/me` itself lands with T4.2.1 (fenced by BLK-1).
 * The wrapper targets the contracted path now; against a server that has not
 * mounted it yet the call is a `404`, which is the correct failure and is not
 * worked around here.
 *
 * The one thing this module adds to the body is `versions.wrapper`, which the
 * server cannot know — it is this package's own published version. It is a
 * **fill, never an overwrite**: a key the server supplied wins, including
 * `wrapper` itself, so a future host that has a better answer is not silently
 * overruled by the client.
 *
 * It is deliberately not a namespace class. `endpoints.json` names this
 * operation `client.me()`, a method on the client itself, so the module exports
 * a function over a narrow host interface and `W6wClient` delegates to it —
 * same layering as the namespaces, without inventing a `client.me.me()`.
 *
 * @module
 */

import { ApiError } from "./errors.ts";
import type { HttpResponse, RequestOptions } from "./http.ts";
import type { Me } from "./types.ts";
import { VERSION } from "./version.ts";

/**
 * The slice of `W6wClient` this operation needs: the transport, and nothing
 * else. Structural rather than a concrete client type, so it stays
 * independently callable in a test and this module never imports the client
 * back.
 */
export interface MeHost {
  /** Perform one request. */
  request<T>(options: RequestOptions): Promise<HttpResponse<T>>;
}

/**
 * Fetch the caller's identity and the component versions behind it.
 *
 * The identity fields are returned exactly as they arrived — unknown ones
 * included, since the return is an interface and not a closed schema.
 * `versions` is normalised to always exist and to always carry a `wrapper`
 * entry, so a caller printing a version banner never has to branch on whether
 * the server sent the block at all.
 *
 * @param host - The client this operation issues its request through.
 * @returns The caller's identity, with `versions.wrapper` filled in.
 * @throws {ApiError} On any non-2xx, e.g. `401` when the token is not accepted.
 * @throws {ApiError} `bad_response` when a 2xx body is not a plain object — an
 * array included, since `typeof [] === "object"` and a spread array yields
 * character-indexed keys rather than identity fields.
 */
export async function fetchMe(host: MeHost): Promise<Me> {
  const res = await host.request<Me>({ method: "GET", path: "/me" });
  const body = res.body;
  // A body that is not an object cannot be identity data. Raise rather than
  // spread it — the same `bad_response` class the envelope reader uses for the
  // keyed routes — so a proxy answering `200 "ok"` fails here instead of
  // returning a `Me` whose only populated field is the one this client just
  // added.
  //
  // `Array.isArray` is load-bearing, not belt-and-braces: `typeof [] === "object"`
  // and `[] !== null`, so without it a `200 []` spread into exactly the degenerate
  // value described above — measured, `{"versions":{"wrapper":"0.1.0"}}` — and
  // `200 [{...}]` produced character-indexed keys. It also keeps this lane in step
  // with python, whose natural `isinstance(body, dict)` excludes a list already.
  if (typeof body !== "object" || body === null || Array.isArray(body)) {
    throw new ApiError(
      res.status,
      "bad_response",
      `Server returned a ${res.status} whose body is not an identity object.`,
      body,
    );
  }
  // Same reasoning one level down: `versions` is typed `Record<string, string>`,
  // but the wire can carry anything, and spreading a string produces
  // character-indexed junk (`"1.2.3"` → `{"0":"1","1":".",…}`) that satisfies the
  // type and so is invisible to the compiler. A `versions` that is not a plain
  // object is dropped in favour of the client's own fill rather than merged —
  // this operation reports identity, and inventing keys out of a malformed block
  // would be worse than reporting only what this client actually knows.
  const versions = body.versions;
  const merged = typeof versions === "object" && versions !== null && !Array.isArray(versions)
    ? versions
    : undefined;
  return {
    ...body,
    // Spread order is the policy: the wrapper's own version is the DEFAULT and
    // every key the server sent lands on top of it, `wrapper` included.
    versions: { wrapper: VERSION, ...merged },
  };
}

File: python/src/w6w/me.py
"""`client.me()` — who the caller is, and what answered.

One GET to `/auth/me` — the server's real identity route (verified live: `200`,
`{tenant, subject, account, role}`). The response body is **flat**, so there is
nothing for `unwrap` to do here; this is one of the two operations that reads a
flat body, and it asserts the body's shape with
:func:`w6w.types.require_object` instead.

**Fixed 2026-07-28** (`docs/endpoints.md` §1): every wrapper calls `/auth/me`
directly rather than waiting on the `/me` alias D15 originally specced and that
was never built. The live consumer of that route is the studio, whose typed `Me`
is the flat body — a nested envelope would break it. What is still missing
server-side is the optional `versions` block, which this module tolerates being
absent entirely.

It is deliberately **not a namespace class**. `endpoints.json` names this
operation `client.me()` — a method on the client itself — so this module exports
a function over a narrow request seam and `Client` delegates to it, rather
than inventing a `client.me.me()`.

── The one value this module adds, and the one it must not ──
`versions["wrapper"]` is filled in here because the server cannot know it: it is
this package's own published version. It is a **default, never an overwrite** —
a key the server supplied wins, `wrapper` included, so a future host with a
better answer is not silently overruled by the client
(`docs/implementation.md` §5, "precedence on collision is server-wins").

The map is otherwise **carried through unaltered**. A server that sends an
unbumped placeholder version reaches the caller with that value intact;
presenting such a value as `dev` is a separate *rendering* rule (D5) that applies
where a version is shown to a person — the CLI's `w6w info` banner. The two rules
act at different layers and this module implements exactly one of them: a wrapper
does not edit the map it returns, because a bug report is read off these values
and it has to quote what the server actually said.
"""

from __future__ import annotations

from dataclasses import replace
from typing import Any, Mapping, Optional, Protocol

from ._http import HttpResponse
from ._version import __version__
from .types import Me, require_object


class MeRequest(Protocol):
    """The single capability this operation is given: send one request.

    A callable rather than an object with a `request` method, for the same
    reason `vars` takes one: there is nothing else this operation may reach.
    `Client.request` satisfies it as a bound method, and so does a three-line
    fake in a test.
    """

    def __call__(
        self,
        method: str,
        path: str,
        query: Optional[Mapping[str, Any]] = None,
        body: Optional[Any] = None,
    ) -> HttpResponse:
        """Perform one request.

        :param method: HTTP method.
        :param path: Base-relative path.
        :param query: Query parameters; `None` values are dropped.
        :param body: Request body, serialised as JSON.
        :returns: The status and parsed body.
        """
        ...  # pragma: no cover - a protocol body is never executed.


def fetch_me(request: MeRequest) -> Me:
    """Fetch the caller's identity and the component versions behind it.

    Reached as `client.me()`; this function is the implementation the client
    delegates to.

    `versions` is normalised to **always exist and always carry `wrapper`**, so
    a caller printing a version banner never has to branch on whether the server
    sent the block at all. Everything else in the block is the server's.

    :param request: The client's bound `request` method.
    :returns: The identity, with `versions["wrapper"]` filled in as a default.
    :raises ConfigError: When no token is configured.
    :raises ApiError: On any non-2xx — e.g. `401` when the token is not
        accepted.
    :raises ApiError: `bad_response` when a `2xx` body is not an object. A list
        included: `200 []` must fail here rather than return an identity whose
        only populated field is the one this function just added (measured in the
        `node` lane, `evals/T2.1.4.eval.md`; `isinstance(body, dict)` is what
        makes this lane get it right for free).
    """
    response = request("GET", "/auth/me")
    identity = Me.from_wire(require_object(response, "an identity object"))
    # Merge order IS the policy, and the two orders are observably different on
    # the one field a bug report is read off: this package's version is the
    # DEFAULT and every key the server sent lands on top of it. The reverse
    # (`{**identity.versions, "wrapper": __version__}`) would make the wrapper's
    # own value win — see `docs/implementation.md` §5, which pins this literal
    # expression rather than leaving it to be decided three times.
    return replace(identity, versions={"wrapper": __version__, **identity.versions})
